mpi4jax/_src/tokenizer.py

```diff
@@ -0,0 +1,154 @@
+        token, *safe_map(read, eqn.invars)
+    )
+    # We use `auto_tokenize` here since the condition function
+    # is forced to only return a boolean value and cannot return
+    # a token.
+    new_cond_fn = lambda token, *args: auto_tokenize(
+        jax.core.jaxpr_as_fun(bind_params["cond_jaxpr"]), token
+    )(*args)
+    bind_params["cond_jaxpr"] = jax.make_jaxpr(new_cond_fn)(
+        token, *safe_map(read, eqn.invars)
+    )
+    # Update bind_params to account for the additional token.
+    ans = eqn.primitive.bind(token, *safe_map(read, eqn.invars), **bind_params)
+    token = ans[0]
+    ans = ans[1:]  # Drop the token.
+    return token, ans
+
+
+def cond_override(read, eqn, token):
+    _, bind_params = eqn.primitive.get_bind_params(eqn.params)
+    branch_jaxprs = []
+    cond_var = eqn.invars[0]
+    other_vars = eqn.invars[1:]
+    for branch in bind_params["branches"]:
+        new_branch_fn = lambda token, *args: _token_forwarding(
+            jax.core.jaxpr_as_fun(branch), token
+        )(*args)
+        new_jaxpr = jax.make_jaxpr(new_branch_fn)(token, *safe_map(read, other_vars))
+        branch_jaxprs.append(new_jaxpr)
+    bind_params["branches"] = tuple(branch_jaxprs)
+    bind_params["linear"] = (False,) + bind_params["linear"]
+    ans = eqn.primitive.bind(
+        read(cond_var), token, *safe_map(read, other_vars), **bind_params
+    )
+    token = ans[0]
+    ans = ans[1:]  # Drop the token.
+    return token, ans
+
+
+def _override_tokens(jaxpr, consts, token, *args):
+    if token is None:  # Create a new token if one is not passed.
+        token = jax.lax.create_token()
+
+    def read(v):
+        if type(v) is jax.core.Literal:
+            return v.val
+        else:
+            return env[v]
+
+    def write(v, val):
+        env[v] = val
+
+    env = {}
+    write(jax.core.unitvar, jax.core.unit)
+    safe_map(write, jaxpr.constvars, consts)
+    safe_map(write, jaxpr.invars, args)
+    for eqn in jaxpr.eqns:
+        # Here, we override the original primatives with our own
+        # token forcing method.
+        if eqn.primitive in token_override_registry:
+            token_override = token_override_registry[eqn.primitive]
+            ans = token_override(
+                safe_map(read, eqn.invars), new_token=token, **eqn.params
+            )
+            # We pass along the new token returned by the above binding
+            token = ans[-1]
+            safe_map(write, eqn.outvars, ans)
+        else:
+            if eqn.primitive is xla.xla_call_p:
+                token, ans = xla_call_overrride(read, eqn, token)
+            elif eqn.primitive is jax.lax.scan_p:
+                token, ans = scan_override(read, eqn, token)
+            elif eqn.primitive is jax.lax.while_p:
+                token, ans = while_override(read, eqn, token)
+            elif eqn.primitive is jax.lax.cond_p:
+                token, ans = cond_override(read, eqn, token)
```

Comment:
Looks like something that could be handled with another registry.

Reply:
or the same, maybe?

Reply:
The signatures are different between these overrides and the other overrides, so I think different registers would be easier.